keps/sig-api-machinery/20190621-apply-scalability.md

```diff
@@ -0,0 +1,154 @@
+
+## Motivation
+
+### The Problems
+
+From our testing, the problems seem to be because of a large number of small object allocations.
+
+#### On list, update, and apply: 
+
+Deserializing managedFields as a deeply nested field in the apiserver allocates many small objects. Objects are deserialized when fetched from etcd.
+The size of managedFields is similar to the size of the object. It increases etcd size, and cache size of the apiserver and clients.
+
+#### On update and apply: 
+
+Converting managedFields from the api representation into the internal representation causes many small object allocations.
+Some structured-merge-diff operations (like compare) also create many small objects that are not needed. Most “field set” operations are not done in place.
+
+### Experiments
+
+The following experiments convinced us that the problem is mostly the allocations of small objects. Latency in etcd because of read/write with bigger objects, and/or data transfer was not a significant issue.
+
+For all these scenarios, we’ve created a deployment with 5,000 pods. Pods that have the managedFields (when server-side apply is enabled) are obviously larger. Here are some numbers for the comparison, with the objects in protobuf format:
+
+- Server-side apply disabled:
+  - A pod is about 8,137 bytes
+  - 5,000 pods is 40,506,271 bytes
+- Server-side apply enabled:
+  - A pod is about 13,164 bytes (~62% larger)
+  - 5,000 pods is 67,182,690 bytes
+
+All the following scenarios have been tested by LISTing all the pods, then running 250 patches per second while still LISTing the pods at the same time.
+
+1. Server-side apply disabled, 
+2. Server-side apply enabled, 
+3. Server-side apply disabled, but with an extra annotation. The annotation is about 5,000 bytes to compensate for the difference in object size with server-side apply.
+4. Server-side apply disabled, with 320 small annotations. Those annotations total about 5,000 bytes to simulate the small object allocation.
+
+The average latency are as follows:
+
+| Scenario | LIST | PATCH | LIST while PATCH |
+|---|---|---|---|
+| 1. |360.4 ms | 8.0 ms | 361.0 ms |
+| 2. | 1294.5 ms | 38.5 ms | 2529.6 ms |
+| 3. | 480.8 ms | 11.3 ms | 500.2 ms |
+| 4. | 1260.1 ms | 9.7 ms | 1503.2 ms |
+
+### Goals
+
+TODO
+
+### Non-Goals
+
+TODO
+
+## Proposal
+
+We have considered a few overlapping solutions:
+
+### Only send managedFields if requested:
+
+Add a query parameter (or some other mechanism) to allow users to request to see the managedFields field in metadata. If the parameter is missing,  strip the field on reads and writes (in responses). The benefit here is we can skip serializing the field most of the time. Most clients won’t and shouldn’t have to understand how to read this field. So making them receive it is a waste of resources, and confusing.
```

Comment:
Since we've tried to avoid sending back parts of objects (instead preferring the transform use case), this has a pretty high bar to clear for design complexity (even if it might be the simplest).

Also, not returning all objects is confusing because PUT semantics should be resetting that field, so PUT would not be correct if it didn't clear managed fields.

Reply:
> Also, not returning all objects is confusing because PUT semantics should be resetting that field, so PUT would not be correct if it didn't clear managed fields.

Because the managed fields can already be dropped by typed client that are not aware of it, we already ignore updates that just completely clear it. This wouldn't really be a new use-case.

Reply:
> Because the managed fields can already be dropped by typed client that are not aware of it, we already ignore updates that just completely clear it. This wouldn't really be a new use-case.

That is very surprising to me.  That is not what PUT means.

Reply:
PUT should replace the object. But, it doesn't make sense to change both managed fields and other things with the same request, since all updates imply changes to managed fields. There definitely needs to be a way to clear the managed fields.